[pvr] adds capability for sorted channel groups

xhaggi opened the pull request:
This will adds the capability to import the sort position of channel groups from the backends.
One known backend which supports sorting of groups is TVHeadend, but i think there are more.

@Jalle19, @opdenkamp mind taking a look.

xhaggi committed:
[pvr] adds capability for sorted channel groups

opdenkamp commented on xbmc/addons/include/xbmc_pvr_types.h:
```diff
@@ -244,6 +244,7 @@ extern "C" {
   {
     char         strGroupName[PVR_ADDON_NAME_STRING_LENGTH]; /*!< @brief (required) name of this channel group */
     bool         bIsRadio;                                   /*!< @brief (required) true if this is a radio channel group, false otherwise. */
+    int          iPosition;                                  /*!< @brief (optional) order position of the group */
```
I think we best use an `unsigned int` instead here, and default to 0 when the add-on doesn't provide this feature (which is what they should be doing with `memset()` before assigning anyway)

then we can use some other sorting inside Kodi if it's set to 0, e.g. sorting by name

xhaggi replied:
agree.

opdenkamp commented on xbmc/pvr/channels/PVRChannelGroups.cpp:
```diff
@@ -35,6 +35,12 @@
 using namespace PVR;
 
+// predicate used for sorting
+bool compGroupPosition(const CPVRChannelGroupPtr &group1, const CPVRChannelGroupPtr &group2)
```
rather use `compareGroupPosition()` instead here, no need to use abbrev. ;-)

xhaggi replied:
sure ;)

opdenkamp commented:
+1 other than the minor comments

ksooo commented:
What about extending PVR_ADDON_CAPABILITIES with something like <code>bool bSupportsSortedChannelGroups</code>?

Additionally/Alternatively, shouldn't we document a value to set in PVR_CHANNEL_GROUP in case sorted channel groups are not supported by the backend?

opdenkamp commented:
no need for an entry in `PVR_ADDON_CAPABILITIES` if we just default to 0 when not supported, but we should indeed document it too

ksooo commented:
I'm fine with this. To document it is sufficient.

xhaggi commented:
I see one issue if we only rely on `iPosition = 0` to indicate that the backend don't support sorting, we need to check all groups if they hold 0 and therefore fallback to another sort logic. In TVH the default sort index is 0 and if we have 2 groups with 0 and one group with 1 it should be sorted by position.

Jalle19 commented on xbmc/pvr/channels/PVRChannelGroups.cpp:
```diff
@@ -102,6 +111,14 @@ bool CPVRChannelGroups::Update(const CPVRChannelGroup &group, bool bUpdateFromCl
+  CSingleLock lock(m_critSection);
+
+  // sort groups by position
+  std::sort(m_groups.begin(), m_groups.end(), compGroupPosition);
```
Since we can use C++11 now there's no need to pollute the namespace with a comparator function.

```
std::sort(m_groups.begin(), m_groups.end(), 
    [](const CPVRChannelGroupPtr &left, const CPVRChannelGroupPtr &right)
{
    return left->GetPosition() < right->GetPosition();
}
```

xhaggi replied:
yep thanks, this was added before we switch to c++11.

xhaggi commented:
after a second look it's better if we check the sort positions instead of introducing a new addon capability because it's possible to have 2 backends with same group names which we merge together in kodi.

addressed all comments and added the check if we need to sort by position. As we currently do not sort groups i didn't add a fallback.

ksooo commented on xbmc/addons/include/xbmc_pvr_types.h:
```diff
@@ -244,6 +244,7 @@ extern "C" {
     char         strGroupName[PVR_ADDON_NAME_STRING_LENGTH]; /*!< @brief (required) name of this channel group */
     bool         bIsRadio;                                   /*!< @brief (required) true if this is a radio channel group, false otherwise. */
+    unsigned int iPosition = 0;                              /*!< @brief (optional) order position of the group */
   } ATTRIBUTE_PACKED PVR_CHANNEL_GROUP;
```
Could you please document in the comment that 0 is to be used to indicate that the backend does not support this feature.

xhaggi replied:
ahh sorry forgot about it ;)

xhaggi commented:
@ksooo done

xhaggi commented:
jenkins build this please
